PackageReport class

jayqi opened the pull request:
New class `PackageReport` that serves as a bundle of individual reporters for a single package. This class has a `render_report` method to render the pkgnet HTML package report. `CreatePackageReport` now uses this under the hood to render the report, and returns an object of this class.

TODO:
- [x] NEWS.md
- [x] Unit test PackageReport class

Resolves #224.

bburns632 commented on R/CreatePackageReport.R:
```diff
@@ -8,122 +265,70 @@
+        X = pkg_reporters
+        , FUN = function(reporter) {
+            class <- class(reporter)[1]
+            createdReport[[class]] <- reporter
```
@jayqi can you explain a little more what's going on here?  Is this syntax to refer to the active methods that are built in `PackageReport`?  What are your thoughts on an `AddReporter` method for `PackageReport`?

jayqi replied:
Yeah, this is how active bindings work. So in general you can do `myObject$myBinding <- 'value'` or whatever. 

That's a good idea about adding a method. I agree that the current method is somewhat clunky for interactive use.

jameslamb commented on R/CreatePackageReport.R:
```diff
@@ -1,3 +1,261 @@
+            }
+            return(private$protected$report_path)
+        }
+        , SummaryReporter = function(reporter) {
```
ugh this is one of those cases where GitHub's PR review is BRUTAL compared to Bitbucket. It's hard for me to see which files got touched.

Here's my question: why was it necessary to have named active bindings for each of these? Since we're just `lapply`-ing over "whatever reporters you give us" (https://github.com/UptakeOpenSource/pkgnet/blob/master/inst/package_report/package_report.Rmd#L31), and you aren't actually doing any special handling here, why is it necessary to have `DependencyReporter`, `FunctionReporter`, etc.

jayqi replied:
So there is special handling, but I ended up moving it all into `private$set_reporter` so I could repeat less code. If we do this with public fields instead, then there's no way to prevent someone from adding a non-reporter object using `<-` assignment right?

jayqi commented on R/CreatePackageReport.R:
```diff
@@ -1,3 +1,261 @@
+        , render_report = function() {
+            log_info("Rendering package report...")
+
+            silence_logger()
```
🤷‍♂ I copied it from the existing code https://github.com/UptakeOpenSource/pkgnet/blob/master/R/CreatePackageReport.R#L91

I'll take a closer look and try to see what it does in this instance, if anything.

jayqi replied:
It seems like this does nothing so I'm getting rid of them.

jayqi commented:
@bburns632 @jameslamb I think this is more or less ready for re-review. I've addressed comments other than one by @jameslamb that I left open for further discussion.

The build is failing for some mysterious reason on the mac OS test. Gonna dig into that more over the weekend. Linux seems to be fine.
